File: webapi-sdk/php/webapi.php
<?php

class UnauthorizedException extends Exception {
}


class InvalidResourceException extends Exception
{
}


class InvalidPlayerException extends Exception
{
}


class InvalidTimeException extends Exception
{
}


class InvalidMethodException extends Exception
{
}


class Status
{
    public $players;
    public $player_limit;
    public $seed;

    public function __construct($result)
    {
        $this->players = $result['players'];
        $this->player_limit = $result['player-limit'];
        $this->seed = $result['seed'];
    }

    public function __toString()
    {
        return sprintf('Status {players=[%s], player_limit=%s, seed=%s}',
                        implode(', ', $this->players), $this->player_limit,
                        $this->seed);
    }
}


class Player
{
    public $has_equipment = false;
    public $has_skills = false;
    public $name;
    public $position;
    public $class_type;
    public $specialization;
    public $level;
    public $power_level;
    public $equipment = array();
    public $skills = array();

    public function __construct($result)
    {
        $this->has_equipment = isset($result['equipment']);
        $this->has_skills = isset($result['skills']);
        $this->name = $result['name'];
        $this->position = array('x' => $result['position']['x'],
                                'y' => $result['position']['y'],
                                'z' => $result['position']['z']);
        $this->class_type = $result['class'];
        $this->specialization = $result['specialization'];
        $this->level = $result['level'];
        $this->power_level = $result['power-level'];
        if($this->has_equipment)
        {
            $this->parse_equipment($result['equipment']);
        }
        if($this->has_skills)
        {
            $this->parse_skills($result['skills']);
        }
    }

    private function parse_equipment($result)
    {
        foreach($result as $item)
        {
            $this->equipment[] = new Item($item);
        }
    }

    private function parse_skills($result)
    {
        $this->skills['pet_master'] = $result['pet-master'];
        $this->skills['riding'] = $result['riding'];
        $this->skills['climbing'] = $result['climbing'];
        $this->skills['hang_gliding'] = $result['hang-gliding'];
        $this->skills['swimming'] = $result['swimming'];
        $this->skills['sailing'] = $result['sailing'];
        $this->skills['class_skill_1'] = $result['class-skill-1'];
        $this->skills['class_skill_2'] = $result['class-skill-2'];
        $this->skills['class_skill_3'] = $result['class-skill-3'];
    }

    public function __toString()
    {
        $string = sprintf('Player {name=%s, ...}', $this->name);
        if($this->has_equipment)
        {
            $string .= ' [Has_Equipment]';
        }
        if($this->has_skills)
        {
            $string .= ' [Has_Skills]';
        }
        return $string;
    }
}


class Item
{
    public $type;
    public $sub_type;
    public $modifier;
    public $minus_modifier;
    public $rarity;
    public $material;
    public $flags;
    public $level;
    public $power_level;
    public $upgrades = array();

    public function __construct($result)
    {
        $this->type = $result['type'];
        $this->sub_type = $result['sub-type'];
        $this->modifier = $result['modifier'];
        $this->minus_modifier = $result['minus-modifier'];
        $this->rarity = $result['rarity'];
        $this->material = $result['material'];
        $this->flags = $result['flags'];
        $this->level = $result['level'];
        $this->power_level = $result['power-level'];
        foreach($result['upgrades'] as $upgrade)
        {
            $this->upgrades[] = new ItemUpgrade($upgrade);
        }
    }

    public function __toString()
    {
        $string = sprintf('Item {type=%s, sub_type=%s, rarity=%s, ...}',
                            $this->type, $this->sub_type, $this->rarity);
        $string .= sprintf(" [%s upgrades]",count($this->upgrades));
        return $string;
    }
}


class ItemUpgrade
{
    public $x;
    public $y;
    public $z;
    public $material;
    public $level;

    public function __construct($result)
    {
        $this->x = $result['x'];
        $this->y = $result['y'];
        $this->z = $result['z'];
        $this->material = $result['material'];
        $this->level = $result['level'];
    }

    public function __toString()
    {
        return sprintf('ItemUpgrade {x=%s, y=%s, z=%s, material=%s, level=%s}',
                        $this->x, $this->y, $this->z, $this->material,
                        $this->level);
    }
}


class WebAPI
{
    const URL_FORMAT = 'http://%s:%s/%s';

    private $host;
    private $port;
    private $key;

    public function __construct($key, $host, $port = 12350)
    {
        $this->key = $key;
        $this->host = $host;
        $this->port = $port;

        if(!extension_loaded('cURL'))
        {
            throw new Exception('WebAPI requires PHP cURL extension to work');
        }
    }

    private function generateUrl($endpoint, $secure, $queryParams = null)
    {
        $url = sprintf(WebAPI::URL_FORMAT, $this->host, $this->port, $endpoint);

        if($secure)
        {
            $url .= sprintf("?key=%s", urlencode($this->key));
        }

        if(!is_null($queryParams) && is_array($queryParams))
        {
            $url .= ($secure ? "&" : "?");
            $url .= http_build_query($queryParams);
        }

        return $url;
    }

    private function send($endpoint = '', $data = null, $secure = true, $queryParams = null)
    {
        $url = $this->generateUrl($endpoint, $secure, $queryParams);
        $handle = curl_init($url);
        curl_setopt($handle, CURLOPT_PORT, $this->port);
        curl_setopt($handle, CURLOPT_HEADER, false);
        curl_setopt($handle, CURLOPT_RETURNTRANSFER, true);
        if(!is_null($data))
        {
            curl_setopt($handle, CURLOPT_POST, true);
            curl_setopt($handle, CURLOPT_POSTFIELDS, $data);
        }       
        $result = curl_exec($handle);
        curl_close($handle);
        return $this->parse($result);
    }

    private function handleError($result)
    {
        $errorCode = $result['error'];
        switch($errorCode)
        {
            case -1:
                throw new UnauthorizedException();
            case -2:
                throw new InvalidResourceException();
            case -3:
                throw new InvalidPlayerException();
            case -4:
                throw new InvalidTimeException();
            case -5:
                throw new InvalidMethodException();
        }
    }

    private function parse($data)
    {
        $result = json_decode($data, true);
        if(isset($result['error']))
        {
            $this->handleError($result);
        }
        return $result;
    }

    public function version()
    {
        $result = $this->send('', null, false);
        return $result['version'];
    }

    public function status()
    {
        $result = $this->send('status');
        return new Status($result);
    }

    public function player($name, $includeEquipment = false, $includeSkills = false)
    {
        $inclusion = array();
        if($includeEquipment)
        {
            $inclusion[] = 'equipment';
        }
        if($includeSkills)
        {
            $inclusion[] = 'skills';
        }
        $params = null;
        if(count($inclusion) > 0)
        {
            $params = array('include', implode(',', $inclusion));
        }
        $result = $this->send(sprintf('player/%s', $name), null, true, $params);
        return new Player($result['player']);
    }

    public function kick($name)
    {
        $result = $this->send(sprintf('kick/%s', $name));
        return isset($result['success']);
    }

    public function time($value = null)
    {
        if(is_null($value)) {
            $result = $this->send('time');
            return $result['time'];
        }
        $parsed = date_parse_from_format('H:i', $value);
        if($parsed['error_count'] != 0) {
            throw new InvalidTimeException();
        }
        $result = $this->send(sprintf('time/%s', $value));
        return isset($result['success']);
    }

    public function message($message, $receiver = null)
    {
        $endpoint = 'message/';
        if(!is_null($receiver))
        {
            $endpoint .= sprintf('%s/', $receiver);
        }
        $result = $this->send($endpoint, $message);
        return isset($result['success']);
    }
}

File: webapi-sdk/python/webapi.py

import json
import time
from urllib import urlencode
import urllib2


URL_FORMAT = 'http://{host}:{port}/{endpoint}'


class UnauthorizedError (Exception):
    pass


class InvalidResourceError (Exception):
    pass


class InvalidPlayerError (Exception):
    pass


class InvalidTimeError (Exception):
    pass


class InvalidMethodError (Exception):
    pass


ERROR_CODES = {
    -1: UnauthorizedError,
    -2: InvalidResourceError,
    -3: InvalidPlayerError,
    -4: InvalidTimeError,
    -5: InvalidMethodError
}


class Status (object):
    def __init__(self, result):
        self.players = result['players']
        self.player_limit = result['player-limit']
        self.seed = result['seed']

    def __str__(self):
        return 'Status {players=%s, player_limit=%s, seed=%s}' % (
            self.players, self.player_limit, self.seed)


class Player (object):
    def __init__(self, result):
        self.has_equipment = 'equipment' in result
        self.has_skills = 'skills' in result
        self.name = result['name']
        self.position = {'x': result['position']['x'],
                         'y': result['position']['y'],
                         'z': result['position']['z']}
        self.class_type = result['class']
        self.specialization = result['specialization']
        self.level = result['level']
        self.power_level = result['power-level']
        if self.has_equipment:
            self._parse_equipment(result['equipment'])
        if self.has_skills:
            self._parse_skills(result['skills'])

    def _parse_equipment(self, result):
        self.equipment = []
        for item in result:
            self.equipment.append(Item(item))

    def _parse_skills(self, result):
        self.skills = {
            'pet_master': result['pet-master'],
            'riding': result['riding'],
            'climbing': result['climbing'],
            'hang_gliding': result['hang-gliding'],
            'swimming': result['swimming'],
            'sailing': result['sailing'],
            'class_skill_1': result['class-skill-1'],
            'class_skill_2': result['class-skill-2'],
            'class_skill_3': result['class-skill-3'],
        }

    def __str__(self):
        string = "Player {name=%s, position=%s, ...}" % (
            self.name, self.position)
        if self.has_equipment:
            string += ' [Has_Equipment]'
        if self.has_skills:
            string += ' [Has_Skills]'
        return string


class Item (object):
    def __init__(self, result):
        self.type = result['type'],
        self.sub_type = result['sub-type']
        self.modifier = result['modifier']
        self.minus_modifier = result['minus-modifier']
        self.rarity = result['rarity']
        self.material = result['material']
        self.flags = result['flags']
        self.level = result['level']
        self.power_level = result['power-level']
        self.upgrades = []
        for upgrade in result['upgrades']:
            self.upgrades.append(ItemUpgrade(upgrade))

    def __str__(self):
        string = "Item {type=%s, sub_type=%s, rarity=%s, ...}" % (
            self.type, self.sub_type, self.rarity)
        string += " [%s upgrades]" % len(self.upgrades)
        return string


class ItemUpgrade (object):
    def __init__(self, result):
        self.x = result['x']
        self.y = result['y']
        self.z = result['z']
        self.material = result['material']
        self.level = result['level']

    def __str__(self):
        return "ItemUpgrade {x=%s, y=%s, z=%s, material=%s, level=%s}" % (
            self.x, self.y, self.z, self.material, self.level)


class WebAPI (object):
    def __init__(self, key, host, port=12350):
        self.__key = key
        self.__host = host
        self.__port = port
        self._setup_opener()

    def _setup_opener(self):
        opener = urllib2.build_opener()
        opener.addheaders = [('User-Agent', 'CuwoAPI/0.0.3')]
        self.opener = opener

    def _generate_url(self, endpoint, secure, query_params=None):
        url = URL_FORMAT.format(host=self.__host, port=self.__port,
                                endpoint=endpoint)
        if secure:
            url += '?' + urlencode({'key': self.__key})

        if query_params is not None:
            if secure:
                url += '&'
            else:
                url += '?'
            params = urlencode(query_params)
            url += params
        return url

    def _send(self, endpoint='', data=None, secure=True, query_params=None):
        url = self._generate_url(endpoint, secure, query_params)
        request = urllib2.Request(url, data)
        if data is not None:
            request.add_header("Content-Type",
                               "application/x-www-form-urlencoded")
        result = self.opener.open(request).read()
        return self._parse(result)

    def _handle_error(self, result):
        error_code = result['error']
        raise ERROR_CODES[error_code]()

    def _parse(self, data):
        result = json.loads(data)
        if 'error' in result:
            self._handle_error(result)
        return result

    def version(self):
        result = self._send(secure=False)
        return result['version']

    def status(self):
        result = self._send('status')
        return Status(result)

    def player(self, name, include_equipment=False, include_skills=False):
        inclusion = []
        if include_equipment:
            inclusion.append('equipment')
        if include_skills:
            inclusion.append('skills')
        params = None
        if(len(inclusion) > 0):
            params = {'include': ','.join(inclusion)}
        result = self._send('player/%s' % name, query_params=params)
        return Player(result['player'])

    def kick(self, name):
        result = self._send('kick/%s' % name)
        return 'success' in result

    def time(self, value=None):
        if value is None:
            result = self._send('time')
            return result['time']
        try:
            time.strptime(value, '%H:%M')
            result = self._send('time/%s' % value)
            return 'success' in result
        except ValueError:
            raise InvalidTimeError()

    def message(self, message, receiver=None):
        endpoint = 'message/'
        if receiver is not None:
            endpoint += '%s/' % receiver
        result = self._send(endpoint, message)
        return 'success' in result


def main():
    api = WebAPI('demo', '127.0.0.1')

    # Get version
    print 'Version: %s' % api.version()

    # Get status
    status = api.status()
    print status

    # Get player details
    if 'Xharon' in status.players:
        player = api.player('Xharon', include_equipment=True,
                            include_skills=True)
        print player

    # Set time
    if api.time('16:05'):
        print 'Time set'

    # Get time
    time = api.time()
    print 'Time: %s' % time

    # Send global message
    if api.message('Hello server - API'):
        print 'Message has been sent'

    # Send message to player
    if 'Xharon' in status.players:
        if api.message('Hey. How are you?', 'Xharon'):
            print 'Message has been sent to Xharon'

    # Kick player
    if 'Xharon' in status.players:
        if api.kick('Xharon'):
            print 'Xharon has been kicked'

if __name__ == '__main__':
    main()
